plugins/outputs/postgresql/utils/utils.go

```diff
@@ -0,0 +1,168 @@
+package utils
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/jackc/pgx"
+)
+
+const (
+	insertIntoSQLTemplate = "INSERT INTO %s(%s) VALUES(%s)"
+)
+
+// GroupMetricsByMeasurement groups the list of metrics by the measurement name.
+// But the values are the index of the measure from the input list of measures.
+// [m, m, m2, m2, m] => {m:[0,1,4], m2:[2,3]}
+func GroupMetricsByMeasurement(m []telegraf.Metric) map[string][]int {
+	toReturn := make(map[string][]int)
+	for i, metric := range m {
+		var metricLocations []int
+		var ok bool
+		name := metric.Name()
+		if metricLocations, ok = toReturn[name]; !ok {
+			metricLocations = []int{}
+			toReturn[name] = metricLocations
+		}
+		toReturn[name] = append(metricLocations, i)
+	}
+	return toReturn
+}
+
+// BuildJsonb returns a byte array of the json representation
+// of the passed object.
+func BuildJsonb(data interface{}) ([]byte, error) {
+	d, err := json.Marshal(data)
+	if err != nil {
+		return nil, err
+	}
+	return d, nil
+}
+
+// QuoteIdent returns a sanitized string safe to use in SQL as an identifier
+func QuoteIdent(name string) string {
+	return pgx.Identifier{name}.Sanitize()
+}
+
+// QuoteLiteral returns a sanitized string safe to use in sql as a string literal
+func QuoteLiteral(name string) string {
+	return "'" + strings.Replace(name, "'", "''", -1) + "'"
+}
+
+// FullTableName returns a sanitized table name with it's schema (if supplied)
+func FullTableName(schema, name string) *pgx.Identifier {
+	if schema != "" {
+		return &pgx.Identifier{schema, name}
+	}
+
+	return &pgx.Identifier{name}
+}
+
+const (
+	PgBool                     = "boolean"
+	PgInt8                     = "int8"
+	PgInt4                     = "int4"
+	PgInteger                  = "integer"
+	PgBigInt                   = "bigint"
+	PgFloat8                   = "float8"
+	PgDoublePrecision          = "double precision"
+	PgText                     = "text"
+	PgTimestamptz              = "timestamptz"
+	PgTimestampWithTimeZone    = "timestamp with time zone"
+	PgTimestamp                = "timestamp"
+	PgTimestampWithoutTimeZone = "timestamp without time zone"
+	PgSerial                   = "serial"
+	PgJSONb                    = "jsonb"
+)
+
+// DerivePgDatatype returns the appropriate PostgreSQL data type
+// that could hold the value.
+func DerivePgDatatype(value interface{}) PgDataType {
+	switch value.(type) {
+	case bool:
+		return PgBool
+	case uint64:
```

Comment:
not strictly needed but since you can’t do fallthrough you can do:
```go
case uint64, int64, int, uint:
  return PgInt8
case uint32, int32:
  return PgInt4
case float64, float32:
  return PgFloat8
```

can ask @iand if that is good style or not

Reply:
I'd prefer it this way